### PR TITLE
Added NMState nodeNetworkState struct

### PR DESCRIPTION
**Pull request by evgenLevin**

### CODE REVIEW

File: pkg/nmstate/nodenetworkstate.go [hunk 1]
```diff
@@ -1 +1,183 @@
...
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// StateBuilder provides struct for the NodeNetworkState object containing connection to
+// the cluster and the NodeNetworkState definitions.
+type StateBuilder struct {
+	Definition *nmstateV1alpha1.NodeNetworkState
```
Comment (kononovn) [inline pkg/nmstate/nodenetworkstate.go:24]: Add comment to public structs

File: pkg/nmstate/nodenetworkstate.go [hunk 2]
```diff
@@ -1 +1,183 @@
...
+	Object     *nmstateV1alpha1.NodeNetworkState
+	apiClient  *clients.Settings
+	errorMsg   string
+}
+
+// NewStateBuilder creates a new instance of StateBuilder.
+func NewStateBuilder(apiClient *clients.Settings, name string) *StateBuilder {
```
Comment (kononovn) [inline pkg/nmstate/nodenetworkstate.go:31]: Do we have an scenario when we need to create NodeNetworkState manually ?

File: pkg/nmstate/nodenetworkstate.go [hunk 3]
```diff
@@ -1 +1,183 @@
...
+	if valid, err := builder.validate(); !valid {
+		return 0, err
+	}
+
+	glog.V(100).Infof(
+		"Getting total-vfs under interface %s from NodeNetworkState %s",
+		sriovInterface, builder.Definition.Name)
+
+	if sriovInterface == "" {
+		glog.V(100).Infof("The sriovInterface  can not be empty string")
```
Comment (kononovn) [inline pkg/nmstate/nodenetworkstate.go:104]: Please remove space.

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix minor comments

File: pkg/nmstate/nodenetworkstate.go (hunk 2 above)
Comment (evgenLevin) [reply pkg/nmstate/nodenetworkstate.go]: removed

**Comment by evgenLevin** [comment]
@kononovn PTAL

File: pkg/nmstate/nodenetworkstate.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// StateBuilder provides struct for the NodeNetworkState object containing connection to
+// the cluster and the NodeNetworkState definitions.
+type StateBuilder struct {
+	// NodeNetworkState definition. Used to create NodeNetworkState object with minimum set of required elements.
+	Definition *nmstateV1alpha1.NodeNetworkState
```
Comment (kononovn) [inline pkg/nmstate/nodenetworkstate.go:25]: This object is created automatically by cluster. Please remove this var.

File: pkg/nmstate/nodenetworkstate.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+// GetTotalVFs returns total-vfs under the given interface.
+func (builder *StateBuilder) GetTotalVFs(sriovInterfaceName string) (int, error) {
+	if valid, err := builder.validate(); !valid {
+		return 0, err
+	}
+
+	glog.V(100).Infof(
+		"Getting total-vfs under interface %s from NodeNetworkState %s",
+		sriovInterfaceName, builder.Definition.Name)
```
Comment (kononovn) [inline pkg/nmstate/nodenetworkstate.go:82]: builder.Object.Name

File: pkg/nmstate/nodenetworkstate.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	}
+
+	var CurrentState DesiredState
+
+	err := yaml.Unmarshal(builder.Object.Status.CurrentState.Raw, &CurrentState)
+	if err != nil {
+		return 0, fmt.Errorf("failed to Unmarshal NMState state")
+	}
+
+	for _, interf := range CurrentState.Interfaces {
```
Comment (kononovn) [inline pkg/nmstate/nodenetworkstate.go:97]: Please re-name to interfaceFromCurrentState

File: pkg/nmstate/nodenetworkstate.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+func (builder *StateBuilder) validate() (bool, error) {
+	resourceCRD := "NodeNetworkState"
+
+	if builder == nil {
+		glog.V(100).Infof("The %s builder is uninitialized", resourceCRD)
+
+		return false, fmt.Errorf("error: received nil %s builder", resourceCRD)
+	}
+
+	if builder.Definition == nil {
```
Comment (kononovn) [inline pkg/nmstate/nodenetworkstate.go:145]: Please remove this check

Comment (kononovn) [reply pkg/nmstate/nodenetworkstate.go:145]: I think you need to change it to builder.Object

**Review by kononovn — CHANGES_REQUESTED** [review]
looks good just few minor comments to fix

File: pkg/nmstate/shared.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package nmstate
+
+// DesiredState provides struct for the NMState desired state object containing all NMState configuration.
+type DesiredState struct {
+	Interfaces []Interface `yaml:"interfaces"`
+}
+
+// Interface provides struct for the NMState interface state object containing interface information.
+type Interface struct {
```
Comment (kononovn) [inline pkg/nmstate/shared.go:9]: Please rename struct to NetworkInterface in order to avoid name overlapping with golang interface

**Review by kononovn — CHANGES_REQUESTED** [review]
1 last comment to fix

File: pkg/nmstate/shared.go (hunk 8 above)
Comment (evgenLevin) [reply pkg/nmstate/shared.go]: renamed

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**